engines/afalg/e_afalg.c

```diff
@@ -230,7 +230,11 @@ int afalg_fin_cipher_aio(afalg_aio *aio, int sfd, unsigned char *buf,
     memset(cb, '\0', sizeof(*cb));
     cb->aio_fildes = sfd;
     cb->aio_lio_opcode = IOCB_CMD_PREAD;
-    cb->aio_buf = (uint64_t)buf;
+    /*
+     * The pointer has to be converted to unsigned value first to avoid
+     * sign extension on cast to 64 bit value
+     */
+    cb->aio_buf = (uint64_t)(unsigned long)buf;
```

Comment:
The more I think about it, using `__x86_64__ && __ILP32__` means the same truncation problem could occur on Aarch32 execution environments on Aarch64.

The core problem seems to be stretching a 32-bit pointer to a 64-bit value due to the kernel. The two data models where that's possible is `ILP32` and `LLP64`. So maybe something like this would be best:

#if __ILP32__ || __LLP64__
cb->aio_buf = (uint64_t)(unsigned long)buf;
#else
cb->aio_buf = (uint64_t)buf;
#endif